pkg/resultstore/resultstore.go

```diff
@@ -166,6 +185,40 @@ type store struct {
 	artifactPubSub   chans.PubSub[ArtifactEvent]
 	artifactSubMutex sync.RWMutex
 	artifactMutex    sync2.KeyedMutex[uint64]
+
+	frozen bool
+}
+
+func (s *store) Freeze() {
```

Comment:
This `Freeze` implementation is interesting. Could be good, but I think the current implementation is overly complex and error prone to race conditions.

This pubsub system isn't that easy to code around. I wonder if we could change it to rely more on file handles instead or something.

So, the flow should something like this:

- worker is done with the build
- aggregator waits until all logs and other data has been transmitted
- aggregator kills works

I'm not a fan of how this feezing logic gets a different code flow and a different freeze-check when starting the stream after the build is done.

We should do a pair-programming session for this and brainstorm some in front of a whiteboard, because I'm having issues wrapping my head around our current solution here.

Reply:
See #82